### PR TITLE
fstar: build with dune

### LINKED ISSUE #275866: fstar: fstarlib / fstar.lib not being installed anymore
### Describe the bug

In the F* package, it appears that `fstarlib` / `fstar.lib` is not being installed anymore:

```bash
$ ls /nix/store/gpfmizgg7d98p5a2c53iii2vlf9dqbng-fstar-2023.09.03/lib/ocaml/4.14.1/site-lib/fstarlib/
total 0
```

This makes it impossible to build programs written in F* and extracted to OCaml.

I suspect this problem was introduced in 20b2d94235fd3319653b48741eaa54b8a9137b41, as I believe in NixOS 23.05 this problem didn't exist.

### Steps To Reproduce

1. Build / install fstar.
2. Look inside the Nix store installation directory to see the `fstarlib` / `fstar.lib` OCaml lib was not installed.

### Expected behavior

I expected the library to be installed.

### Notify maintainers

cc @marsam @gebner @pnmadelaine 

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by pnmadelaine**
## Description of changes

Build F* with Dune, reflecting changes to the F* build process from a few months ago.
This splits the build into three derivations, one for the F* binary, one for the library and one for the complete package.

Should fix #275866 

cc @W95Psp who helped rework the F* flake, does this look good to you?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by someplaceguy** [comment]
@pnmadelaine Thanks for fixing this!

I've just tried this PR and F* builds correctly for me and seems to work fine.

However, I am running into one issue. I have a tiny F* program which gets extracted to OCaml and builds fine when using PR #275889 but fails to build with this PR unless I apply the following patch to my program's `default.nix`:

```diff
diff --git a/default.nix b/default.nix
index 7a8e9c7..49b7c49 100644
--- a/default.nix
+++ b/default.nix
@@ -25,6 +25,18 @@ stdenv.mkDerivation {

   nativeBuildInputs = with ocamlPackages; [ findlib fstar ocaml ocamlbuild tup z3_4_8_5 ];

+  buildInputs = with ocamlPackages; [
+    batteries
+    menhirLib
+    pprint
+    process
+    ppx_deriving
+    ppx_deriving_yojson
+    sedlex
+    stdint
+    zarith
+  ];
+
   installPhase = ''
     mkdir -p $out/bin
```

The issue is that `fstar.lib` requires those libraries (but my program doesn't).

This is normally fixed by moving these libraries in the `fstarDune` derivation from `buildInputs` to `propagatedBuildInputs` so that downstream derivations don't have to declare these dependencies manually.

However, since in this case the `fstarDune` derivation is not the actual derivation that gets exported to `all-packages.nix`, it might be slightly more awkward. I think it would require not only moving those libraries to `propagatedBuildInputs` in `fstarDune`, but it might also require adding `fstarDune` to the `buildInputs` (or `propagatedBuildInputs`?) of the main `fstar` derivation as well, I think...

**Comment by pnmadelaine** [comment]
@someplaceguy thanks for testing this!
I don't think adding `fstarDune` as a propagated build input is a good idea, since it would expose two F* binaries.
Instead I propagated `fstarDune.buildInputs` directly.
It does not feel too awkard to me, what do you think?

**Comment by pnmadelaine** [comment]
Just noticed some things in `fstarDune.buildInputs` are not necessary to propagate.
I am just going to do the same split you made in your PR if that's okay!

**Comment by someplaceguy** [comment]
>  It does not feel too awkard to me, what do you think?

I agree it's not too awkward. I just tried your latest changes (including the `don't propagate everything` commit) and it seems to work perfectly.

Many thanks! :tada:

File: pkgs/development/compilers/fstar/dune.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+, yojson
+, zarith
+}:
+
+buildDunePackage {
+  pname = "fstar";
+  inherit version src;
+
+  duneVersion = "3";
+
```
Comment (vbgl) [inline pkgs/development/compilers/fstar/dune.nix:24]:
```suggestion
```

This line is not needed.

**Review by vbgl — APPROVED** [review]

**Pull request merged**